File: examples/python/hand/agg/handlers/state.py
"""Hand aggregate state management."""

import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

from google.protobuf.any_pb2 import Any

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "angzarr"))
sys.path.insert(0, str(Path(__file__).parent.parent))

from angzarr_client.proto.angzarr import types_pb2 as types
from angzarr_client.proto.examples import hand_pb2 as hand
from angzarr_client.proto.examples import types_pb2 as poker_types


@dataclass
class PlayerHandInfo:
    """State for a player in the hand."""

    player_root: bytes = b""
    position: int = 0
    hole_cards: list = field(default_factory=list)  # List of (suit, rank) tuples
    stack: int = 0
    bet_this_round: int = 0
    total_invested: int = 0
    has_acted: bool = False
    has_folded: bool = False
    is_all_in: bool = False


@dataclass
class PotInfo:
    """State for a pot."""

    amount: int = 0
    eligible_players: list = field(default_factory=list)  # List of player_root bytes
    pot_type: str = "main"


@dataclass
class HandState:
    """Complete state for a hand aggregate."""

    hand_id: str = ""
    table_root: bytes = b""
    hand_number: int = 0
    game_variant: int = 0

    # Deck state - list of (suit, rank) tuples
    remaining_deck: list = field(default_factory=list)

    # Player state - position -> PlayerHandInfo
    players: dict = field(default_factory=dict)

    # Community cards - list of (suit, rank) tuples
    community_cards: list = field(default_factory=list)

    # Betting state
    current_phase: int = 0  # BettingPhase enum
    action_on_position: int = -1
    current_bet: int = 0
    min_raise: int = 0
    pots: list = field(default_factory=list)  # List of PotInfo

    # Positions
    dealer_position: int = 0
    small_blind_position: int = 0
    big_blind_position: int = 0
    small_blind: int = 0
    big_blind: int = 0

    status: str = ""  # "dealing", "betting", "showdown", "complete"

    def exists(self) -> bool:
        """Check if hand has been dealt."""
        return self.status != ""

    def get_active_players(self) -> list:
        """Get players still in the hand (not folded, not all-in)."""
        return [
            p for p in self.players.values() if not p.has_folded and not p.is_all_in
        ]

    def get_players_in_hand(self) -> list:
        """Get all players still in hand (not folded)."""
        return [p for p in self.players.values() if not p.has_folded]

    def get_next_to_act(self) -> Optional[int]:
        """Find next position that needs to act."""
        active = self.get_active_players()
        if len(active) <= 1:
            return None

        positions = sorted(p.position for p in active)
        if not positions:
            return None

        # Find next position after current
        for pos in positions:
            if pos > self.action_on_position:
                player = self.players.get(pos)
                if player and not player.has_acted:
                    return pos

        # Wrap around
        for pos in positions:
            player = self.players.get(pos)
            if player and not player.has_acted:
                return pos

        return None

    def is_betting_complete(self) -> bool:
        """Check if current betting round is complete."""
        active = self.get_active_players()
        if len(active) <= 1:
            return True

        # All active players must have acted and matched current bet
        for player in active:
            if not player.has_acted:
                return False
            if player.bet_this_round < self.current_bet and not player.is_all_in:
                return False

        return True

    def get_pot_total(self) -> int:
        """Get total pot amount."""
        return sum(p.amount for p in self.pots)


def rebuild_state(event_book: types.EventBook) -> HandState:
    """Rebuild hand state from event history."""
    state = HandState()

    if not event_book.pages:
        return state

    # Sort pages by sequence number to ensure correct order
    def get_seq(p):
        if p.WhichOneof('sequence') == 'num':
            return p.num
        return 0
    sorted_pages = sorted(event_book.pages, key=get_seq)

    for page in sorted_pages:
        event_any = page.event
        type_url = event_any.type_url

        # NOTE: Check CommunityCardsDealt BEFORE CardsDealt since "CardsDealt" is a substring
        if "CommunityCardsDealt" in type_url:
            event = hand.CommunityCardsDealt()
            event_any.Unpack(event)

            for card in event.cards:
                dealt_card = (card.suit, card.rank)
                state.community_cards.append(dealt_card)
                # Remove from remaining deck
                if dealt_card in state.remaining_deck:
                    state.remaining_deck.remove(dealt_card)

            state.current_phase = event.phase
            state.status = "betting"  # Ready for next betting round

            # Reset for next betting round
            for player in state.players.values():
                player.bet_this_round = 0
                player.has_acted = False
            state.current_bet = 0

        elif "CardsDealt" in type_url:
            event = hand.CardsDealt()
            event_any.Unpack(event)
            state.hand_id = f"{event.table_root.hex()}_{event.hand_number}"
            state.table_root = event.table_root
            state.hand_number = event.hand_number
            state.game_variant = event.game_variant
            state.dealer_position = event.dealer_position
            state.status = "betting"
            state.current_phase = poker_types.PREFLOP

            # Initialize players
            for player in event.players:
                state.players[player.position] = PlayerHandInfo(
                    player_root=player.player_root,
                    position=player.position,
                    stack=player.stack,
                )

            # Collect all dealt cards
            dealt_cards = set()

            # Set hole cards
            for pc in event.player_cards:
                if pc.player_root in [p.player_root for p in state.players.values()]:
                    for pos, player in state.players.items():
                        if player.player_root == pc.player_root:
                            player.hole_cards = [(c.suit, c.rank) for c in pc.cards]
                            for c in pc.cards:
                                dealt_cards.add((c.suit, c.rank))

            # Reconstruct remaining deck (all cards not dealt)
            full_deck = []
            for suit in [
                poker_types.CLUBS,
                poker_types.DIAMONDS,
                poker_types.HEARTS,
                poker_types.SPADES,
            ]:
                for rank in range(2, 15):
                    card = (suit, rank)
                    if card not in dealt_cards:
                        full_deck.append(card)
            # Shuffle the remaining deck to randomize community card order
            import random
            random.shuffle(full_deck)
            state.remaining_deck = full_deck

            # Initialize main pot
            state.pots = [
                PotInfo(
                    amount=0,
                    eligible_players=[p.player_root for p in state.players.values()],
                    pot_type="main",
                )
            ]

        elif "BlindPosted" in type_url:
            event = hand.BlindPosted()
            event_any.Unpack(event)

            for player in state.players.values():
                if player.player_root == event.player_root:
                    player.stack = event.player_stack
                    player.bet_this_round = event.amount
                    player.total_invested += event.amount
                    if event.blind_type == "small":
                        state.small_blind_position = player.position
                        state.small_blind = event.amount
                    elif event.blind_type == "big":
                        state.big_blind_position = player.position
                        state.big_blind = event.amount
                        state.current_bet = event.amount
                        state.min_raise = event.amount
                    break

            if state.pots:
                state.pots[0].amount = event.pot_total

            state.status = "betting"

        elif "ActionTaken" in type_url:
            event = hand.ActionTaken()
            event_any.Unpack(event)

            for player in state.players.values():
                if player.player_root == event.player_root:
                    player.stack = event.player_stack
                    player.has_acted = True

                    if event.action == poker_types.FOLD:
                        player.has_folded = True
                    elif event.action in (
                        poker_types.CALL,
                        poker_types.BET,
                        poker_types.RAISE,
                    ):
                        player.bet_this_round += event.amount
                        player.total_invested += event.amount
                    elif event.action == poker_types.ALL_IN:
                        player.is_all_in = True
                        player.bet_this_round += event.amount
                        player.total_invested += event.amount

                    if event.action in (
                        poker_types.BET,
                        poker_types.RAISE,
                        poker_types.ALL_IN,
                    ):
                        if player.bet_this_round > state.current_bet:
                            raise_amount = player.bet_this_round - state.current_bet
                            state.current_bet = player.bet_this_round
                            state.min_raise = max(state.min_raise, raise_amount)
                    break

            if state.pots:
                state.pots[0].amount = event.pot_total

            state.action_on_position = -1  # Will be set by process manager

        elif "BettingRoundComplete" in type_url:
            event = hand.BettingRoundComplete()
            event_any.Unpack(event)

            # Reset for next round
            for player in state.players.values():
                player.bet_this_round = 0
                player.has_acted = False

            state.current_bet = 0
            if state.pots:
                state.pots[0].amount = event.pot_total

            # Advance phase for draw-based games (Five Card Draw)
            # Community card games advance phase when cards are dealt
            from .game_rules import get_game_rules
            rules = get_game_rules(state.game_variant)
            if rules.variant == poker_types.FIVE_CARD_DRAW:
                transition = rules.get_next_phase(event.completed_phase)
                if transition:
                    state.current_phase = transition.next_phase

        elif "DrawCompleted" in type_url:
            event = hand.DrawCompleted()
            event_any.Unpack(event)

            for player in state.players.values():
                if player.player_root == event.player_root:
                    # Remove discarded cards and add new ones
                    # For now, just update with new cards
                    player.hole_cards = [(c.suit, c.rank) for c in event.new_cards]
                    break

        elif "ShowdownStarted" in type_url:
            state.status = "showdown"

        elif "CardsRevealed" in type_url:
            pass  # Just for display

        elif "PotAwarded" in type_url:
            event = hand.PotAwarded()
            event_any.Unpack(event)

            for winner in event.winners:
                for player in state.players.values():
                    if player.player_root == winner.player_root:
                        player.stack += winner.amount
                        break

        elif "HandComplete" in type_url:
            state.status = "complete"

    return state

File: examples/go/hand/agg/handlers/state.go
// Package handlers implements hand aggregate state reconstruction.
package handlers

import (
	"encoding/hex"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
)

// HandState represents the current state of a hand aggregate.
type HandState struct {
	HandID      string
	TableRoot   []byte
	HandNumber  int64
	GameVariant examples.GameVariant

	// Deck state
	RemainingDeck []*examples.Card

	// Player state
	Players map[string]*PlayerHandState // player_root_hex -> state

	// Community cards
	CommunityCards []*examples.Card

	// Betting state
	CurrentPhase     examples.BettingPhase
	ActionOnPosition int32
	CurrentBet       int64
	MinRaise         int64
	Pots             []*PotState

	// Positions
	DealerPosition     int32
	SmallBlindPosition int32
	BigBlindPosition   int32

	Status string // "dealing", "betting", "showdown", "complete"
}

// PlayerHandState represents a player's state in the hand.
type PlayerHandState struct {
	PlayerRoot    []byte
	Position      int32
	HoleCards     []*examples.Card
	Stack         int64
	BetThisRound  int64
	TotalInvested int64
	HasActed      bool
	HasFolded     bool
	IsAllIn       bool
}

// PotState represents a pot (main or side).
type PotState struct {
	Amount          int64
	EligiblePlayers [][]byte
	PotType         string
}

// NewHandState creates a new empty hand state.
func NewHandState() HandState {
	return HandState{
		Players: make(map[string]*PlayerHandState),
		Pots:    []*PotState{{PotType: "main"}},
	}
}

// Exists returns true if the hand has been dealt.
func (s HandState) Exists() bool {
	return s.HandID != ""
}

// IsComplete returns true if the hand is complete.
func (s HandState) IsComplete() bool {
	return s.Status == "complete"
}

// TotalPot returns the sum of all pots.
func (s HandState) TotalPot() int64 {
	total := int64(0)
	for _, pot := range s.Pots {
		total += pot.Amount
	}
	return total
}

// GetPlayerByRoot returns the player state for a given player root.
func (s HandState) GetPlayerByRoot(root []byte) *PlayerHandState {
	return s.Players[hex.EncodeToString(root)]
}

// ActivePlayerCount returns the number of players who haven't folded.
func (s HandState) ActivePlayerCount() int {
	count := 0
	for _, p := range s.Players {
		if !p.HasFolded {
			count++
		}
	}
	return count
}

// Event applier functions for StateRouter

func applyCardsDealt(state *HandState, event *examples.CardsDealt) {
	state.HandID = hex.EncodeToString(event.TableRoot) + "_" + string(rune(event.HandNumber))
	state.TableRoot = event.TableRoot
	state.HandNumber = event.HandNumber
	state.GameVariant = event.GameVariant
	state.DealerPosition = event.DealerPosition
	state.RemainingDeck = event.RemainingDeck
	state.CurrentPhase = examples.BettingPhase_PREFLOP
	state.Status = "betting"

	// Initialize players
	for _, p := range event.Players {
		key := hex.EncodeToString(p.PlayerRoot)
		state.Players[key] = &PlayerHandState{
			PlayerRoot: p.PlayerRoot,
			Position:   p.Position,
			Stack:      p.Stack,
		}
	}

	// Apply hole cards
	for _, pc := range event.PlayerCards {
		key := hex.EncodeToString(pc.PlayerRoot)
		if player := state.Players[key]; player != nil {
			player.HoleCards = pc.Cards
		}
	}
}

func applyBlindPosted(state *HandState, event *examples.BlindPosted) {
	key := hex.EncodeToString(event.PlayerRoot)
	if player := state.Players[key]; player != nil {
		player.Stack = event.PlayerStack
		player.BetThisRound += event.Amount
		player.TotalInvested += event.Amount
	}
	state.Pots[0].Amount = event.PotTotal
	if event.Amount > state.CurrentBet {
		state.CurrentBet = event.Amount
	}
}

func applyActionTaken(state *HandState, event *examples.ActionTaken) {
	key := hex.EncodeToString(event.PlayerRoot)
	if player := state.Players[key]; player != nil {
		player.Stack = event.PlayerStack
		player.HasActed = true

		switch event.Action {
		case examples.ActionType_FOLD:
			player.HasFolded = true
		case examples.ActionType_ALL_IN:
			player.IsAllIn = true
			player.BetThisRound += event.Amount
			player.TotalInvested += event.Amount
		case examples.ActionType_BET, examples.ActionType_RAISE, examples.ActionType_CALL:
			player.BetThisRound += event.Amount
			player.TotalInvested += event.Amount
		}
	}
	state.Pots[0].Amount = event.PotTotal
	state.CurrentBet = event.AmountToCall
}

func applyBettingRoundComplete(state *HandState, event *examples.BettingRoundComplete) {
	// Reset for next round
	for _, p := range state.Players {
		p.BetThisRound = 0
		p.HasActed = false
	}
	state.CurrentBet = 0

	// Update stacks from snapshot
	for _, snap := range event.Stacks {
		key := hex.EncodeToString(snap.PlayerRoot)
		if player := state.Players[key]; player != nil {
			player.Stack = snap.Stack
			player.IsAllIn = snap.IsAllIn
			player.HasFolded = snap.HasFolded
		}
	}
}

func applyCommunityCardsDealt(state *HandState, event *examples.CommunityCardsDealt) {
	state.CommunityCards = event.AllCommunityCards
	state.CurrentPhase = event.Phase
}

func applyDrawCompleted(state *HandState, event *examples.DrawCompleted) {
	key := hex.EncodeToString(event.PlayerRoot)
	if player := state.Players[key]; player != nil {
		// Replace discarded cards with new cards
		if len(event.NewCards) > 0 {
			player.HoleCards = append(player.HoleCards[:len(player.HoleCards)-int(event.CardsDiscarded)], event.NewCards...)
		}
	}
	// Update remaining deck
	if int(event.CardsDrawn) <= len(state.RemainingDeck) {
		state.RemainingDeck = state.RemainingDeck[event.CardsDrawn:]
	}
}

func applyShowdownStarted(state *HandState, _ *examples.ShowdownStarted) {
	state.Status = "showdown"
}

func applyCardsRevealed(state *HandState, _ *examples.CardsRevealed) {
	// Cards revealed during showdown - could store revealed hands
}

func applyCardsMucked(state *HandState, _ *examples.CardsMucked) {
	// Player mucked - could mark as mucked
}

func applyPotAwarded(state *HandState, event *examples.PotAwarded) {
	for _, winner := range event.Winners {
		key := hex.EncodeToString(winner.PlayerRoot)
		if player := state.Players[key]; player != nil {
			player.Stack += winner.Amount
		}
	}
}

func applyHandComplete(state *HandState, event *examples.HandComplete) {
	state.Status = "complete"
	// Update final stacks
	for _, snap := range event.FinalStacks {
		key := hex.EncodeToString(snap.PlayerRoot)
		if player := state.Players[key]; player != nil {
			player.Stack = snap.Stack
		}
	}
}

// stateRouter is the fluent state reconstruction router.
var stateRouter = angzarr.NewStateRouter(NewHandState).
	On(applyCardsDealt).
	On(applyBlindPosted).
	On(applyActionTaken).
	On(applyBettingRoundComplete).
	On(applyCommunityCardsDealt).
	On(applyDrawCompleted).
	On(applyShowdownStarted).
	On(applyCardsRevealed).
	On(applyCardsMucked).
	On(applyPotAwarded).
	On(applyHandComplete)

// RebuildState rebuilds hand state from event history.
func RebuildState(eventBook *pb.EventBook) HandState {
	if eventBook == nil {
		return NewHandState()
	}

	state := NewHandState()
	for _, page := range eventBook.Pages {
		event := page.GetEvent()
		if event != nil {
			stateRouter.ApplySingle(&state, event)
		}
	}
	return state
}
